### PR TITLE
ZWave implement protection command class

### PR DESCRIPTION
**Pull request by jongj**
including converter, channel definition and unit tests.

Example  thing channel config:  
```
     <channel id="protection_local" typeId="protection_local">
        <label>Local Protection</label>
        <properties>
          <property name="binding:*:DecimalType">PROTECTION;type=PROTECTION_LOCAL</property>
        </properties>
      </channel>
      <channel id="protection_rf" typeId="protection_rf">
        <label>Rf Protection</label>
        <properties>
          <property name="binding:*:DecimalType">PROTECTION;type=PROTECTION_RF</property>
        </properties>
      </channel>
 ```
Where protection_rf is only supported by CC version 2 

Signed-off-by: Jorg de Jong <jorg@dejong.info> (github: jongj)

### CODE REVIEW

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        Enum<?> e = (Enum<?>) event.getValue();
+        return new DecimalType(e.ordinal());
+    }
+
+    @Override
+    public List<SerialMessage> receiveCommand(ZWaveThingChannel channel, ZWaveNode node, Command command) {
+        String type = channel.getArguments().get("type");
+
+        ZWaveProtectionCommandClass commandClass = (ZWaveProtectionCommandClass) node
+                .resolveCommandClass(ZWaveCommandClass.CommandClass.PROTECTION, channel.getEndpoint());
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java]: There probably should be null protection here?

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java [hunk 2]
```diff
@@ -64,10 +111,243 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
         logger.debug("NODE {}: Received protection command (v{})", this.getNode().getNodeId(), this.getVersion());
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: Can you change this to 
```
NODE {}: Received PROTECTION command V{}
```

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java [hunk 3]
```diff
@@ -64,10 +111,243 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
...
+        PROTECTION_RF
+    }
+
+    /**
+     * Z-Wave LocalProtectionType enumeration. The LocalProtection type indicates the type of local protection that is
+     * reported.
+     *
+     * @author Jorg de Jong
+     */
+    @XStreamAlias("localProtection")
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: Do we need these 'complex' enums? The label isn't used anywhere (is it) so why not just use a standard enum without the label?
I also wonder if the values shouldn't be uppercase...

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java [hunk 4]
```diff
@@ -64,10 +111,243 @@ public void handleApplicationCommandRequest(SerialMessage serialMessage, int off
         logger.debug("NODE {}: Received protection command (v{})", this.getNode().getNodeId(), this.getVersion());
         int command = serialMessage.getMessagePayloadByte(offset);
         switch (command) {
+            case PROTECTION_REPORT:
+                int localMode = serialMessage.getMessagePayloadByte(offset + 1) & 0x0f;
+
+                if (localMode < LocalProtectionType.values().length) {
+                    this.currentLocalMode = LocalProtectionType.values()[localMode];
```
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: I don't think 'this' is needed when referencing currentLocalMode? I would prefer to not use this unless it's needed.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java (hunk 1 above)
Comment (jongj) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java]: Dou you mean a check on the commandClass?

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java (hunk 2 above)
Comment (jongj) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: done

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java (hunk 3 above)
Comment (jongj) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: Happy to change it and remove the label. Camel case enums seems to be inline with other enums uses in the binding ie ZWaveThingChannel.DataType. But again happy to change.

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java (hunk 4 above)
Comment (jongj) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: removed and also other occurrences.

**Comment by jongj** [comment]
Updated after review:
- added null check
- simplified enums and upcased them
- removed 'this' in CC

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java (hunk 1 above)
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/converter/ZWaveProtectionConverter.java]: I think resolveCommandClass can return null if the class isn't found so we probably should check for this with -:
```
if(commandClass == null) {
    return null;
}
```

File: addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java (hunk 3 above)
Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: Ok, let's change the enum to remove the label and leave the camel case.  At some stage I might refactor the binding to change this, but you are right that at hte moment camel case is consistent.

Comment (cdjackson) [inline addons/binding/org.openhab.binding.zwave/src/main/java/org/openhab/binding/zwave/internal/protocol/commandclass/ZWaveProtectionCommandClass.java]: And I just see you've done this already :)

Thanks!

**Pull request merged**